### PR TITLE
Reduce Flashing & Optimize ASTableNode/ASCollectionNode Data Loading

### PR DESCRIPTION
**Pull request by Adlai-Holler**
Issuing updates to `UICollectionView` is one of the most expensive operations we perform.

This diff combines the section-delete, section-insert, and first-batch-of-rows-insert operations during initial data load & reload data. This will fix flashing in cases where the first batch of items covers the visible area, and it will reduce the total number of UIKit update operations we perform.

### CODE REVIEW

File: AsyncDisplayKit/Details/ASDataController.mm [hunk 1]
```diff
@@ -140,6 +140,11 @@ - (void)batchLayoutNodesFromContexts:(NSArray<ASIndexedNodeContext *> *)contexts
 #if AS_MEASURE_AVOIDED_DATACONTROLLER_WORK
     [ASDataController _expectToInsertNodes:contexts.count];
 #endif
+  
+  if (contexts.count == 0) {
+    batchCompletionHandler(@[], @[]);
```
Comment (appleguy) [inline AsyncDisplayKit/Details/ASDataController.mm]: Is a comment justified for this one?  I assume this is like zero insertions or deletions, and ultimately calls back to...which code?

If you'd rather the reader follow the code, I trust your judgement.  Just a bit mysterious in a diff review that it needs an early return handler.

**Comment by Adlai-Holler** [comment]
To summarize the nature of this change:

**Comment by Adlai-Holler** [comment]
@nguyenhuy cc #2536 not resolved but relevant. Could you give this a post-commit review?

@garrettmoon Can we get this a lodestone pass when it gets synced to Pinterest?

**Comment by nguyenhuy** [comment]
@Adlai-Holler: Awesome PR! It's a much simpler approach to fix flashes. 

I think we can improve this technique to cover cases in which the visible patch is not the first patch. We can queue up all laid out nodes and insert them all at once when the visible patch is ready. What do you think?

**Comment by Adlai-Holler** [comment]
@nguyenhuy I think that is the ideal approach, but we have an architectural issue to resolve before then – how, given a user that implements UICollectionViewLayout, do we know where that visible patch is?

That is, if the user doesn't get a chance to layout nodes until the nodes are measured and inserted (since UICollectionViewLayout talks directly to UICollectionView), how can we wait until they are laid out to insert them? And how can we let the user drive the measurement process instead of us measuring batches of nodes and THEN handing them to the user to layout?

One good solution might be to create a new protocol such as `ASCollectionNodeLayout` that resembles `UICollectionViewLayout`. The user provides us this object, and we talk to it, and we provide our own `UICollectionViewLayout` that vends data buffered from their async layout.

This new async layout would break the circular dependency (insertion [depends on] layout [depends on] insertion). That is, the async layout would find out about new items before the UICollectionView does. It would stay in sync with the data source, which would be another huge win for reliability, since layouts often talk directly to delegates (such as `collectionView:layout:sizeForItemAtIndexPath`). This async layout object (which may well have to run on the main thread, we'll see) will be responsible for allocating & measuring nodes only as it needs to (which is a huge win for the photo grid use case). 

It would be easy enough for us to ask this new async layout object for `layoutAttributesForElementsInRect:` given the screen + some buffer in case they're scrolling. We can't do this today, because the elements aren't yet inserted in the collection view.

Thoughts @nguyenhuy?

**Comment by nguyenhuy** [comment]
@Adlai-Holler: Yeah, great question on determining visible nodes. I was thinking maybe we can rely on the previous visible paths and assume that the new ones are at least very close. It's definitely the case if items have the same size. I can put up a short working prototype with that assumption.

That being said, your idea is more generalized and should cover all cases. But it looks like a big change. After a discussion with @garrettmoon, @appleguy and @maicki, we agreed that we don't have time to make it to 2.0. @garrettmoon will move #2612 to later milestones.

**Pull request merged**